### DISCUSSION TITLE
How to pass Arguments (like an id) using Next.js

### DISCUSSION THREAD

**Discussion by taylorhlabs** [vercel/swr#864] [Q&A] [answered]
I'm using SWR with Next.js app and I'm stumbling over myself with a specific SWR usage.

I have a page with a bunch of "resource" elements and the user can click on each element to trigger a modal. In the modal, I'd like to show comments related to the clicked resource. This means I need to somehow pass a resource ID to my API so I can only get comments linked to the clicked resource. I'm hoping that with SWR I can cache the comments for each resource, so if a user opens Resource X modal and closes it, then opens Resource Y modal and closes it, then opens Resource X modal again, the comments are already cached for Resource X. My attempted implementation is below. 

1) Ideally, I want to just pass the resourceId directly using my `useSWR` function but I couldn't get that to work. 🙄   Is it possible to pass arguments (like an ID) using SWR? **I read through all the docs, including the one on Arguments on Vercel's site, but I can't get it to work.**
2) Since I couldn't pass the argument directly in the `useSWR` function, I added an `axios` function below it to call the API with the necessary resourceId. My understanding is that the SWR function should create a cache for each call using the props.resourceId as a key and call the getResourceComments function to get the data. However, the getResourceComments function never runs in my implementation.

**How can I use SWR and pass a variable (like resourceId in this case)? Thanks for any input!**

Note that since I'm using Next.js, I don't have a `fetcher` function set up. SWR just works. Maybe that's part of my issue...
```
import useSWR from 'swr';
function ModalCommentsList(props) {
  // for fetching comments (use SWR to handle caching)
  console.log(`In component. resourceId: ${props.resourceId}`); // this prints correct resourceId
  const { data, error } = useSWR([ '/api/get-resource-comments', props.resourceId ], getResourceComments);

  // get comments data for modal since can't seem to pass variable using SWR
  const getResourceComments = async () => {
    try {
      const body = props.resourceId;
      const res = await axios.post('get-resource-comments', body);
      const comments = await res.data;
      console.log(comments);
      }
      } catch (error) {
        console.error(error);
      }
  };

  if (error) return <div>An error occured.</div>;
  if (!data) return <div>Loading ...</div>;

  // list of resources
  return (
    <div className="mb-4 thh-comments-list">
      {data.map((comment) => (
        <div key={comment.id} className="bg-light p-3 rounded mb-3">
          <div>{comment.text}</div>
          <div className="d-flex small text-muted">
            <div className="mr-3">Created: {comment.createdAt}</div>
            <div>Updated: {comment.updatedAt}</div>
          </div>
        </div>
      ))}
    </div>
  );
}
export default ModalCommentsList;
```

**Comment by taylorhlabs** [comment, marked as answer]
I ended up going with the code below. Maybe my stumbling will help someone else. I think I misunderstood the docs originally and overcomplicated it...
```
const fetcher = async (url, resourceId) =>
    await fetch(url, {
      headers: {
        resourceid: resourceId
      }
    }).then((response) => response.json());

  const { data, error } = useSWR([ '/api/get-resource-comments', props.resourceId ], fetcher);
```

**Comment by x-tropy** [comment]
For example if you want to handle URL: `http://localhost:5432/project/a277c2`

`a277c2` is the argument you wants to pass to server and make query.

Then create a folder called `project` under `pages` folder, and create a file called `[id].js` under the `project` folder to handle all project requests.

Inside `[id].js`, use a `getServerSideProps` function:

```
export const getServerSideProps = async context => {
  const projectId = context.params.id
  const result = await.prisma.project.findFirst({
    //...
  })
  return {
    props: {result}
  }
}
```

More information about `getServerSideProps`: [https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props)
